src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.c

```diff
@@ -612,6 +667,17 @@ void CryptoNative_SslCtxSetAlpnSelectCb(SSL_CTX* ctx, SslCtxSetAlpnCallback cb,
 #endif
 }
 
+int32_t CryptoNative_SslSetData(SSL* ssl, void *ptr)
+{
+    return SSL_set_ex_data(ssl, 0, ptr);
+}
+
+void* CryptoNative_SslGetData(SSL* ssl)
+{
+//    void* data = SSL_get_ex_data(ssl, 0, ptr);
+    return SSL_get_ex_data(ssl, 0);
+}
```

Comment:
I spent a lot of time trying to figure out how this API is supposed to be used... and I can't quite figure it out, but I'm pretty sure this is wrong :smile:.

I /think/ that slot 0 in ex_data is pre-wired to know that it's storing an `SSL*`, so if we called `SSL_dup` on this `SSL*` then as part of duplicating the EX_DATA portion it'll call `SSL_dup` on the value stored in this slot... which is bad if it's a GCHandle.

It might be OK to use slot 13 (CRYPTO_EX_INDEX_APP), but I'm not sure.  Every usage of this that I see within OpenSSL code always calls `SSL_get_ex_new_index` first, to associate an allocator/deallocator/duplicator with what is going to go into the slot.  Then they get back a number and need to remember that as their slot ID.

I think we should strive to avoid using the EX_DATA, since it's not easily understood.

Reply:
I don't think that is true. AFAIK the index 0 is for user data.
This is why
```c
/* compatibility */
# define SSL_set_app_data(s,arg)         (SSL_set_ex_data(s,0,(char *)(arg)))
# define SSL_get_app_data(s)             (SSL_get_ex_data(s,0))
```
 exist in the header and why OpenSSL does not use it. This is standard way to associate `void*` state with SSL object.
I certainly used it this way before. Now, if you need more than that, you can register additional slots either for international features for for different parts of your application. Since we do not expose `SSL` and we are the only users I think the use in the PR is safe. I can search for examples but I think other applications are using it in similar way. 

I don't think we use SSL_dup but I get the idea. We only need this for very short period of time before OpenSSL process ClientHello message. I can rest it to NULL in the callback as well as we can do after the handshake is done to avoid dangling pointers. AFIK the selection callback would be called once at most for each SSL. 

Let me know if there are any tests or scenarios would want me to check @bartonjs.

While I still feel storing in SSL is ok and cheap we can take different route. Instead of caching SSL_CONTEXT for each protocol version, we could make the ALPN part of the key. That means we would need to compare the list for each SslStream. That would mean more entries in `contexts` dictionary and more work for each connection.

Reply:
I added reset of the pointer in the ALPN callback. I can possibly also update the function animus to match the SSL_get_app_data/SSL_set_app_data from ssl.h

Reply:
FWIW, it looks like SSL_get_app_data was a holdover from some previous versions, and it has finally been removed in 3.0.

But, I agree that in 1.1 they both had the CRYPTO_EX_INDEX_APP index and didn't use it for SSL_get_app_data... so it's probably OK.